pkg/query/build_tree.go

```diff
@@ -33,27 +41,82 @@ func BuildIteratorFromSchema(fullSchema *schema.Schema, definitionName string, r
 	for _, caveat := range builder.collectedCaveats {
 		result = NewCaveatIterator(result, caveat)
 	}
+
+	// Note: RecursiveIterator wrapping happens at the recursion point,
+	// not at the top level. So we shouldn't have any sentinels left here.
+	if len(builder.recursiveSentinels) > 0 {
+		// This would be an error - sentinels should have been wrapped already
+		return nil, spiceerrors.MustBugf("unwrapped sentinels remaining: %d", len(builder.recursiveSentinels))
+	}
+
 	return result, nil
 }
 
 func (b *iteratorBuilder) buildIteratorFromSchemaInternal(definitionName string, relationName string, withSubRelations bool) (Iterator, error) {
 	id := fmt.Sprintf("%s#%s:%v", definitionName, relationName, withSubRelations)
-	if b.seen[id] {
-		return nil, errors.New("recursive schema iterators are as yet unsupported")
+	// Also check the opposite withSubRelations flag to catch recursion across flags
+	oppositeID := fmt.Sprintf("%s#%s:%v", definitionName, relationName, !withSubRelations)
+
+	// Check if we're currently building this (true recursion)
+	// Check both with the same flag and opposite flag, since recursion can cross the boundary
+	if b.building[id] || b.building[oppositeID] {
```

Comment:
Is this what you were referring to that would change under mutual recursion?

Reply:
Yep!